server/status.go

```diff
@@ -155,6 +165,65 @@ func (s *statusServer) handleLocalStatus(w http.ResponseWriter, r *http.Request,
 	w.Write(b)
 }
 
+// handleLocalLogs handles GET requests for list of available logs.
+func (s *statusServer) handleLocalLogs(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	log.Flush()
+	logFiles, err := log.ListLogFiles()
+	if err != nil {
+		log.Error(err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	b, contentType, err := util.MarshalResponse(r, logFiles, []util.EncodingType{util.JSONEncoding})
+	if err != nil {
+		log.Error(err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("Content-Type", contentType)
+	w.Write(b)
+}
+
+// handleLocalLog handles GET requests for a single log. If no filename is
+// available, it returns 404. The log contents are returned in structured
+// format as JSON.
+func (s *statusServer) handleLocalLog(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	log.Flush()
+	file := ps.ByName("file")
+	reader, err := log.GetLogReader(file)
+	if reader == nil || err != nil {
+		log.Errorf("unable to open log file %s: %s", file, err)
+		http.NotFound(w, r)
+		return
+	}
+	defer reader.Close()
+
+	entry := proto.LogEntry{}
+	var entries []proto.LogEntry
+	decoder := log.NewLogEntryDecoder(reader)
+	for {
+		if err := decoder.Decode(&entry); err != nil {
+			if err == io.EOF {
+				break
+			}
+			log.Error(err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		entries = append(entries, entry)
+	}
+
+	b, contentType, err := util.MarshalResponse(r, entries, []util.EncodingType{util.JSONEncoding})
```

Comment:
We use this everywhere. If it's really a problem, it should be fixed at the MarshalResponse level.